### PR TITLE
Fix the memory leak in RhythmGame. Fixes #295

### LINKED ISSUE #295: RhythmGame leaks sound assets
SoundRecording::loadFromAssets opens an AAset but does not close it.

https://github.com/google/oboe/blob/master/samples/RhythmGame/src/main/cpp/audio/SoundRecording.cpp

The SoundRecording uses the buffer that seems to be owned by the Asset.
This may be OK for the RhythmGame because the assets are opened once and then used for the duration of the app. But an app that needs to load and unload assets might inherit a memory leak if they use the code without modification.

### PR DESCRIPTION
**Pull request by dturner**
As per discussion earlier, I've decoupled the source from the player. Feedback welcome.

### CODE REVIEW

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * limitations under the License.
+ */
+
+
+#include <utils/logging.h>
+#include "AAssetDataSource.h"
+
+
+AAssetDataSource* AAssetDataSource::newFromAssetManager(AAssetManager *assetManager,
+														const char *filename,
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: Indented with tabs!

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+														const int32_t channelCount) {
+
+	// Load the backing track
+	AAsset* asset = AAssetManager_open(assetManager, filename, AASSET_MODE_BUFFER);
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: This whole file seems to have tabs. Four spaces are preferred.

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+	if (asset == nullptr){
+		LOGE("Failed to open track, filename %s", filename);
+		return nullptr;
+	}
+
+	// Get the length of the track (we assume it is stereo 48kHz)
+	off_t trackLength = AAsset_getLength(asset);
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: I think this is in bytes. Might avoid some bugs by calling this trackSizeBytes.

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+	if (audioBuffer == nullptr){
+		LOGE("Could not get buffer for track");
+		return nullptr;
+	}
+
+	// There are 4 bytes per frame because
+	// each sample is 2 bytes and
+	// it's a stereo recording which has 2 samples per frame.
+	int32_t numFrames = static_cast<int32_t>(trackLength / 4);
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: Change 4 to sizeof(int16_t) * channelCount

File: samples/RhythmGame/src/main/cpp/audio/Player.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        for (int i = 0; i < numFrames; ++i) {
+            for (int j = 0; j < mChannelCount; ++j) {
+                targetData[(i*mChannelCount)+j] = data[(mReadFrameIndex*mChannelCount)+j];
+            }
+
+            // Increment and handle wraparound
+            if (++mReadFrameIndex >= totalFrames) mReadFrameIndex = 0;
+        }
+
+    } else {
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/Player.cpp]: Should silence any frames that were not rendered, either the whole or partial buffer.

File: samples/RhythmGame/src/main/cpp/audio/Player.h [hunk 6]
```diff
@@ -27,29 +27,36 @@
...
-class SoundRecording : public RenderableAudio{
+class Player : public RenderableAudio{
 
 public:
-    SoundRecording(const int16_t *sourceData, int32_t numFrames)
-            : mData(sourceData)
-            , mTotalFrames(numFrames)
+    /**
+     * Construct a new Player from the given DataSource. Note that the Player does *not* take
+     * ownership of the source so it's up to the caller to delete the source after it's no longer
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/Player.h]: Document that Multiple players can share the same DataSource

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp (hunk 2 above)
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: That's weird. In both Android Studio and CLion I have my editor set to 4 spaces. Fixed now.

File: samples/RhythmGame/src/main/cpp/Game.cpp [hunk 7]
```diff
@@ -76,6 +90,18 @@ void Game::start() {
     }
 }
 
+void Game::stop(){
+
+    if (mAudioStream != nullptr){
+        mAudioStream->close();
+    }
+
+    delete mClap;
```
[the remark this answers is not in the document]
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/Game.cpp]: Good tip. Refactored to use `std::unique_ptr`.

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp (hunk 1 above)
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: Fixed

File: samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp (hunk 3 above)
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/audio/AAssetDataSource.cpp]: Done

File: samples/RhythmGame/src/main/cpp/audio/Player.cpp (hunk 5 above)
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/audio/Player.cpp]: Done

File: samples/RhythmGame/src/main/cpp/audio/Player.h (hunk 6 above)
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/audio/Player.h]: Done

File: samples/RhythmGame/src/main/cpp/Game.cpp [hunk 8]
```diff
@@ -94,10 +94,9 @@ void Game::stop(){
 
     if (mAudioStream != nullptr){
         mAudioStream->close();
+        mAudioStream = nullptr;
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/Game.cpp]: Add delete mAudioStream;

This doc probably needs updating!
https://github.com/google/oboe/blob/master/docs/FullGuide.md#closing-an-audio-stream

File: samples/RhythmGame/src/main/cpp/audio/Player.cpp [hunk 9]
```diff
@@ -19,30 +19,41 @@
 
 void Player::renderAudio(int16_t *targetData, int32_t numFrames){
 
+    const int32_t channelCount = mSource.getChannelCount();
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/audio/Player.cpp]: Do we guarantee somehow that the AAudio stream channel count matches the channel count of the source files?

Comment (dturner) [reply samples/RhythmGame/src/main/cpp/audio/Player.cpp]: We don't. However, this is documented here: https://github.com/google/oboe/tree/master/samples/RhythmGame#caveats-and-limitations

File: samples/RhythmGame/src/main/cpp/Game.cpp (hunk 8 above)
Comment (dturner) [reply samples/RhythmGame/src/main/cpp/Game.cpp]: There must be a better way than this. Maybe for Oboe v2.0 we should change `openStream` to return a smart pointer where `stream->close()` is called when the stream is destroyed.

Comment (dturner) [reply samples/RhythmGame/src/main/cpp/Game.cpp]: Anyway - done :)

**Pull request merged**